[DOCS] Document Model Optimizer Python API (#14380)
openvinotoolkit/openvino: docs/MO_DG/prepare_model/MO_Python_API.md

## Model Optimizer Python API {#openvino_docs_MO_DG_Python_API}

Model Optimizer (MO) has a Python API for model conversion, which is represented by the `convert_model()` method in the openvino.tools.mo namespace.

>   `convert_model()` has all the functionality available from the command-line tool, plus the ability to pass Python objects (models, extensions) to `convert_model()` directly from memory without leaving the training environment (Jupyter Notebook or training scripts).

Reviewer: ```suggestion
  `convert_model()` has all the functionality available from the command-line tool, plus the ability to pass Python model objects like Pytorch model or TensorFlow Keras model directly without saving it into files and without leaving the training environment (Jupyter Notebook or training scripts). As well as the input models consumed directly from Python, `convert_model` can take OpenVINO extension objects constructed directly in Python for easier conversion of operations that are not supported in OpenVINO (TODO: insert the link).
```

```sh
from openvino.tools.mo import convert_model

ov_model = convert_model("resnet.onnx")
```

MO Python API allows the conversion of PyTorch models.

```sh
import torchvision

model = torchvision.models.resnet50(pretrained=True)
ov_model = convert_model(model, input_shape=[1,3,100,100])
```

The following types are supported as an input model for `convert_model()`:

* PyTorch - `torch.nn.Module`, `torch.jit.ScriptModule`, `torch.jit.ScriptFunction`. Refer to the [Converting a PyTorch Model](@ref openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_PyTorch) article for more details.
* TensorFlow/ TensorFlow2 / Keras - `tf.keras.Model`, `tf.keras.layers.Layer`, `tf.compat.v1.GraphDef`, `tf.Module`, `tf.compat.v1.wrap_function`, `tf.compat.v1.session`, `tf.train.checkpoint`, `tf.python.training.tracking.base.Trackable`(with limitations). Refer to the [Converting a TensorFlow Model](@ref openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_TensorFlow) article for more details.

`convert_model()` accepts all parameters available in the MO command-line tool. Parameters can be specified by Python classes or string analogs, similar to the command-line tool.

> Example 1:

Reviewer: So, to understand how to use it I need first to learn all existing parameters in Model Optimizer and then apply "conversion from the command line to Python" to learn how to use it in Python. Some time ago we discussed the possibility to provide both command line and Python documentation along all the topics explained in MO section like we have for C++ and Python: via tabs. Probably there is no much time to achieve it, but could you estimate it at least? It would be more aligned with approach "Python first".

Reviewer: Great, let's create a ticket for that.

Author: Created the ticket: CVS-103302

# Model Optimizer Python API {#openvino_docs_MO_DG_Python_API}

@sphinxdirective

Model Optimizer (MO) has a Python API for model conversion, which is represented by the ``convert_model()`` method in the openvino.tools.mo namespace.

  ``convert_model()`` has all the functionality available from the command-line tool, plus the ability to pass Python model objects, such as a Pytorch model or TensorFlow Keras model directly, without saving them into files and without leaving the training environment (Jupyter Notebook or training scripts). In addition to input models consumed directly from Python, ``convert_model`` can take OpenVINO extension objects constructed directly in Python for easier conversion of operations that are not supported in OpenVINO.
  ``convert_model()`` returns an openvino.runtime.Model object which can be compiled and inferred or serialized to IR.

> Example of conversion of a PyTorch model directly from memory:

Reviewer 2: ```suggestion
Example of converting a PyTorch model directly from memory:
```

.. code-block:: python

   import torchvision
   
   model = torchvision.models.resnet50(pretrained=True)
   ov_model = convert_model(model)

The following types are supported as an input model for ``convert_model()``:

* PyTorch - ``torch.nn.Module``, ``torch.jit.ScriptModule``, ``torch.jit.ScriptFunction``. Refer to the :doc:`Converting a PyTorch Model<openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_PyTorch>` article for more details.
* TensorFlow / TensorFlow 2 / Keras - ``tf.keras.Model``, ``tf.keras.layers.Layer``, ``tf.compat.v1.Graph``, ``tf.compat.v1.GraphDef``, ``tf.Module``, ``tf.function``, ``tf.compat.v1.session``, ``tf.train.checkpoint``. Refer to the :doc:`Converting a TensorFlow Model<openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_TensorFlow>` article for more details.

``convert_model()`` accepts all parameters available in the MO command-line tool. Parameters can be specified by Python classes or string analogs, similar to the command-line tool.

> Example of usage of native Python classes for setting ``input_shape``, ``mean_values`` and ``layout``:

Reviewer 2: ```suggestion
Example of using native Python classes to set ``input_shape``, ``mean_values`` and ``layout``:
```

.. code-block:: python

   from openvino.runtime import PartialShape, Layout
   
   ov_model = convert_model(model, input_shape=PartialShape([1,3,100,100]), mean_values=[127, 127, 127], layout=Layout("NCHW"))

> Example of usage strings for setting ``input_shape``, ``mean_values`` and ``layout``:

Reviewer 2: ```suggestion
Example of using strings for setting ``input_shape``, ``mean_values`` and ``layout``:
```

.. code-block:: python

   ov_model = convert_model(model, input_shape="[1,3,100,100]", mean_values="[127,127,127]", layout="NCHW")

The ``input`` parameter can be set by a ``tuple`` with a name, shape, and type. The input name of the type string is required in the tuple. The shape and type are optional.
The shape can be a ``list`` or ``tuple`` of dimensions (``int`` or ``openvino.runtime.Dimension``), or ``openvino.runtime.PartialShape``, or ``openvino.runtime.Shape``. The type can be of numpy type or ``openvino.runtime.Type``.

> Example of usage of a tuple in ``input`` parameter for cutting the model:

Reviewer 2: ```suggestion
Example of using a tuple in the ``input`` parameter to cut a model:
```

.. code-block:: python

   ov_model = convert_model(model, input=("input_name", [3], np.float32))

For complex cases, when a value needs to be set in the ``input`` parameter, the ``InputCutInfo`` class can be used. ``InputCutInfo`` accepts four parameters: ``name``, ``shape``, ``type``, and ``value``. 

``InputCutInfo("input_name", [3], np.float32, [0.5, 2.1, 3.4])`` is equivalent of ``InputCutInfo(name="input_name", shape=[3], type=np.float32, value=[0.5, 2.1, 3.4])``.

Supported types for ``InputCutInfo``:

* name: ``string``.
* shape: ``list`` or ``tuple`` of dimensions (``int`` or ``openvino.runtime.Dimension``), ``openvino.runtime.PartialShape``, ``openvino.runtime.Shape``.
* type: ``numpy type``, ``openvino.runtime.Type``.
* value: ``numpy.ndarray``, ``list`` of numeric values, ``bool``.

> Example of usage of ``InputCutInfo`` for freezing an input with value:

Reviewer 2: ```suggestion
Example of using ``InputCutInfo`` to freeze an input with value:
```

.. code-block:: python

   from openvino.tools.mo import convert_model, InputCutInfo
   
   ov_model = convert_model(model, input=InputCutInfo("input_name", [3], np.float32, [0.5, 2.1, 3.4]))

To set parameters for models with multiple inputs, use ``list`` of parameters.
Parameters supporting ``list``: 

* input
* input_shape
* layout 
* source_layout
* dest_layout
* mean_values
* scale_values

> Example of usage of lists for setting shapes, types and layout for multiple inputs:

Reviewer 2: ```suggestion
Example of using lists to set shapes, types and layout for multiple inputs:
```

.. code-block:: python

   ov_model = convert_model(model, input=[("input1", [1,3,100,100], np.float32), ("input2", [1,3,100,100], np.float32)], layout=[Layout("NCHW"), LayoutMap("NCHW", "NHWC")])

``layout``, ``source_layout`` and ``dest_layout`` accept an ``openvino.runtime.Layout`` object or ``string``.

> Example of usage of ``Layout`` class for setting of layout of model input:

Reviewer 2: ```suggestion
Example of using the ``Layout`` class to set the layout of a model input:
```

.. code-block:: python

   from openvino.runtime import Layout
   from openvino.tools.mo import convert_model
   
   ov_model = convert_model(model, source_layout=Layout("NCHW"))

To set both source and destination layouts in the ``layout`` parameter, use the ``LayoutMap`` class. ``LayoutMap`` accepts two parameters: ``source_layout`` and ``target_layout``.

``LayoutMap("NCHW", "NHWC")`` is equivalent to ``LayoutMap(source_layout="NCHW", target_layout="NHWC")``.

> Example of usage of ``LayoutMap`` class for changing of layout of model input:

Reviewer 2: ```suggestion
Example of using the ``LayoutMap`` class to change the layout of a model input:
```

| torch.nn.Module torch.jit.ScriptModule torch.jit.ScriptFunction  | tf.keras.Model tf.keras.layers.Layer tf.compat.v1.GraphDef tf.Module tf.compat.v1.wrap_function tf.compat.v1.session tf.train.checkpoint tf.python.training.tracking.base.Trackable(with limitations) | 

- convert_model() accepts all parameters available in MO command line tool. Parameters can be specified by Python classes or string analogs identically to command line tool.
- Example 1:
```sh
from openvino.runtime import PartialShape, Layout

ov_model = convert_model(model, input_shape=PartialShape([1,3,100,100]), mean_values=[127, 127, 127], layout=Layout("NCHW"))
```
- Example 2:
```sh
ov_model = convert_model(model, input_shape="[1,3,100,100]", mean_values="[127,127,127]", layout="NCHW")
```
- Command-line flags like --compress_to_fp16 can be set in Python API by providing a boolean value (True or False).
```sh
ov_model = convert_model(model, compress_to_fp16=True)
```

> - "input" parameter can be set by tuple with name, shape and type. Another option is to use InputCutInfo class which was introduced for complex cases when value also needs to be set.

Reviewer 3: is there any discription InputCutInfo in docs? For now, it is not obvious in which cases user needs to provide value and in which format (e.g. is it always list or maybe numpy or tensor)

Reviewer 3: and type of input, also is always numpy?

Author: Added more info on InputCutInfo and supported types.

- Example 1:
```sh
ov_model = convert_model(model, input=("input_name", [3], np.float32))
```
- Example 2:
```sh
from openvino.tools.mo import convert_model, InputCutInfo

ov_model = convert_model(model, input=InputCutInfo("input_name", [3], np.float32, [0.5, 2.1, 3.4]))
```

> - "layout", "source_layout" and "dest_layout" accept openvino.runtime.Layout object. To set both source and destination layouts LayoutMap class can be used.

Reviewer 3: can you please add example of usae Layout object too? And also maybe provide some comments regarding LayoutMap about arguments order (where source layout and where destination?)

Author: Done.

> ## Model Optimizer Python API {#openvino_docs_MO_DG_prepare_model_MO_Python_API}

Reviewer 4: I would suggest to have openvino_docs_MO_DG_MO_Python_API or openvino_docs_MO_DG_Python_API

Author: Changed to openvino_docs_MO_DG_Python_API.

- Model Optimizer has Python API for model conversion, which is represented by convert_model() method in openvino.tools.mo namespace.

>   convert_model() has all the functionality available from command line tool plus the possibility of passing Python objects (models, extensions) to convert_model() directly from memory.

Reviewer 4: convert_model() has all the functionality available from command line tool plus the possibility of passing Python objects (models, extensions) to convert_model() directly from memory without leaving the training environment (Jupyter Notebook or training scripts).

  convert_model() returns openvino.runtime.Model object which can be compiled and infered or serialized to IR.

```sh
from openvino.tools.mo import convert_model

ov_model = convert_model("resnet.onnx")
```

MO Python API allows conversion of PyTorch models.
```sh
import torchvision

model = torchvision.models.resnet50(pretrained=True)
ov_model = convert_model(model, input_shape=[1,3,100,100])
```

- Following types are supported as input model for convert_model():

| PyTorch                                                          |                                                                                            TF/ TF2 / Keras                                                                                            |
|:-----------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|

> | torch.nn.Module torch.jit.ScriptModule torch.jit.ScriptFunction  | tf.keras.Model tf.keras.layers.Layer tf.compat.v1.GraphDef tf.Module tf.compat.v1.wrap_function tf.compat.v1.session tf.train.checkpoint tf.python.training.tracking.base.Trackable(with limitations) | 

Reviewer 4: I think the table is not the best choise to list of all supported model objects. I think you need to have a couple of bullets for PyTorch and TensorFlow and encounter them. Also in the each bullet please refer to FW specific chapters for more details.

Author: Changed to bullets and added links to corresponding pages. Please check.

- convert_model() accepts all parameters available in MO command line tool. Parameters can be specified by Python classes or string analogs identically to command line tool.
- Example 1:

> ```sh

Reviewer 4: Should not we have blank line between code and the text line? How is it rendered?

Reviewer 4: please add blank line

Author: I checked html generated by Documentation CI job and rendering looks fine. I added blank lines anyway just in case.

## Model Optimizer Python API {#openvino_docs_MO_DG_Python_API}

> - Model Optimizer has Python API for model conversion, which is represented by convert_model() method in openvino.tools.mo namespace.

Reviewer 4: Should we need to use `convert_model()`?

Author: Done.

- Example 1:

```sh
from openvino.runtime import PartialShape, Layout

ov_model = convert_model(model, input_shape=PartialShape([1,3,100,100]), mean_values=[127, 127, 127], layout=Layout("NCHW"))
```

- Example 2:

```sh
ov_model = convert_model(model, input_shape="[1,3,100,100]", mean_values="[127,127,127]", layout="NCHW")
```

- Command-line flags like `--compress_to_fp16` can be set in Python API by providing a boolean value (`True` or `False`).

```sh
ov_model = convert_model(model, compress_to_fp16=True)
```

- `input` parameter can be set by `tuple` with name, shape and type. Input name of type string is required in such tuple. Shape and type are optional. 
Shape can be a `list` or `tuple` of dimensions (`int` or `openvino.runtime.Dimension`) or `openvino.runtime.PartialShape` or `openvino.runtime.Shape`. Type can be of numpy type or `openvino.runtime.Type`.

```sh
ov_model = convert_model(model, input=("input_name", [3], np.float32))
```

- For complex cases when value needs to be set in `input` parameter `InputCutInfo` class can be used. `InputCutInfo` accepts four parameters: `name`, `shape`, `type`, `value`.

>   `InputCutInfo("input_name", [3], np.float32, [0.5, 2.1, 3.4])` is equivalent of `InputCutInfo(name="input_name", shape=[3], type=np.float32, value=[0.5, 2.1, 3.4])`.

Author: Added a new line. Please take a look.

Model Optimizer (MO) has a Python API for model conversion, which is represented by the `convert_model()` method in the openvino.tools.mo namespace.
  `convert_model()` has all the functionality available from the command-line tool, plus the ability to pass Python objects (models, extensions) to `convert_model()` directly from memory without leaving the training environment (Jupyter Notebook or training scripts).
  `convert_model()` returns an openvino.runtime.Model object which can be compiled and inferred or serialized to IR.

```sh
from openvino.tools.mo import convert_model

ov_model = convert_model("resnet.onnx")
```

MO Python API allows the conversion of PyTorch models.

```sh
import torchvision

model = torchvision.models.resnet50(pretrained=True)
ov_model = convert_model(model, input_shape=[1,3,100,100])
```

The following types are supported as an input model for `convert_model()`:

* PyTorch - `torch.nn.Module`, `torch.jit.ScriptModule`, `torch.jit.ScriptFunction`. Refer to the [Converting a PyTorch Model](@ref openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_PyTorch) article for more details.

> * TensorFlow/ TensorFlow2 / Keras - `tf.keras.Model`, `tf.keras.layers.Layer`, `tf.compat.v1.GraphDef`, `tf.Module`, `tf.compat.v1.wrap_function`, `tf.compat.v1.session`, `tf.train.checkpoint`, `tf.python.training.tracking.base.Trackable`(with limitations). Refer to the [Converting a TensorFlow Model](@ref openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_TensorFlow) article for more details.

Reviewer: Please understand the current speed of conversion for objects loaded directly from Python. It turns out there is a regression in the speed comparing to earlier commits in master branch making this feature unusable. If degradation is confirmed, please remove this stuff from documentation.

Author: For 2022.3 branch I added documentation only for phase 1 due to regression:  https://github.com/openvinotoolkit/openvino/pull/14696

Phase 2 documentation in this PR will go to next release.

`convert_model()` accepts all parameters available in the MO command-line tool. Parameters can be specified by Python classes or string analogs, similar to the command-line tool.
Example 1:

```sh
from openvino.runtime import PartialShape, Layout

ov_model = convert_model(model, input_shape=PartialShape([1,3,100,100]), mean_values=[127, 127, 127], layout=Layout("NCHW"))
```

Example 2:

```sh
ov_model = convert_model(model, input_shape="[1,3,100,100]", mean_values="[127,127,127]", layout="NCHW")
```

Command-line flags, like `--compress_to_fp16`, can be set in the Python API by providing a boolean value (`True` or `False`).

```sh
ov_model = convert_model(model, compress_to_fp16=True)
```

The `input` parameter can be set by a `tuple` with a name, shape, and type. The input name of the type string is required in the tuple. The shape and type are optional.
The shape can be a `list` or `tuple` of dimensions (`int` or `openvino.runtime.Dimension`), or `openvino.runtime.PartialShape`, or `openvino.runtime.Shape`. The type can be of numpy type or `openvino.runtime.Type`.

```sh
ov_model = convert_model(model, input=("input_name", [3], np.float32))
```

> For complex cases, when a value needs to be set in the `input` parameter, the `InputCutInfo` class can be used. `InputCutInfo` accepts four parameters: `name`, `shape`, `type`, and `value`. 

Reviewer 5: Could you please provide examples for multi-input cases? For example, I have model with 2 inputs, and I need to set shape for them

Author: Added example of multi-input case. Please check.

# Model Optimizer Python API {#openvino_docs_MO_DG_Python_API}

@sphinxdirective

Model Optimizer (MO) has a Python API for model conversion, which is represented by the ``convert_model()`` method in the openvino.tools.mo namespace.

>   ``convert_model()`` has all the functionality available from the command-line tool, plus the ability to pass Python model objects like Pytorch model or TensorFlow Keras model directly without saving it into files and without leaving the training environment (Jupyter Notebook or training scripts). As well as the input models consumed directly from Python, ``convert_model`` can take OpenVINO extension objects constructed directly in Python for easier conversion of operations that are not supported in OpenVINO.

Author: Done.

   
   model = torchvision.models.resnet50(pretrained=True)
   ov_model = convert_model(model)

The following types are supported as an input model for ``convert_model()``:

* PyTorch - ``torch.nn.Module``, ``torch.jit.ScriptModule``, ``torch.jit.ScriptFunction``. Refer to the :doc:`Converting a PyTorch Model<openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_PyTorch>` article for more details.
* TensorFlow/ TensorFlow2 / Keras - ``tf.keras.Model``, ``tf.keras.layers.Layer``, ``tf.compat.v1.Graph``, ``tf.compat.v1.GraphDef``, ``tf.Module``, ``tf.function``, ``tf.compat.v1.session``, ``tf.train.checkpoint``. Refer to the :doc:`Converting a TensorFlow Model<openvino_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_TensorFlow>` article for more details.

``convert_model()`` accepts all parameters available in the MO command-line tool. Parameters can be specified by Python classes or string analogs, similar to the command-line tool.

Example 1:

.. code-block:: python

   from openvino.runtime import PartialShape, Layout
   
   ov_model = convert_model(model, input_shape=PartialShape([1,3,100,100]), mean_values=[127, 127, 127], layout=Layout("NCHW"))

> Example 2:

Reviewer 4: Should not we tell what this example demonstrating? The same comment for other examples.

Author: Added comments for all examples.

The shape can be a ``list`` or ``tuple`` of dimensions (``int`` or ``openvino.runtime.Dimension``), or ``openvino.runtime.PartialShape``, or ``openvino.runtime.Shape``. The type can be of numpy type or ``openvino.runtime.Type``.

.. code-block:: python

   ov_model = convert_model(model, input=("input_name", [3], np.float32))

For complex cases, when a value needs to be set in the ``input`` parameter, the ``InputCutInfo`` class can be used. ``InputCutInfo`` accepts four parameters: ``name``, ``shape``, ``type``, and ``value``. 

``InputCutInfo("input_name", [3], np.float32, [0.5, 2.1, 3.4])`` is equivalent of ``InputCutInfo(name="input_name", shape=[3], type=np.float32, value=[0.5, 2.1, 3.4])``.

Supported types for ``InputCutInfo``:

* name: ``string``.
* shape: ``list`` or ``tuple`` of dimensions (``int`` or ``openvino.runtime.Dimension``), ``openvino.runtime.PartialShape``, ``openvino.runtime.Shape``.
* type: ``numpy type``, ``openvino.runtime.Type``.
* value: ``numpy.ndarray``, ``list`` of numeric values, ``bool``.

> .. code-block:: python

Reviewer 4: Should we tell that this example demonstrating freezing `input_name` input to some value?
Also, is the command below is valid without listing other inputs that are not frozen?

Author: Added comments for all examples.

   from openvino.tools.mo import convert_model, InputCutInfo
   
   ov_model = convert_model(model, input=InputCutInfo("input_name", [3], np.float32, [0.5, 2.1, 3.4]))

To set parameters for models with multiple inputs, use ``list`` of parameters.
Parameters supporting ``list``: 

* input
* input_shape
* layout 
* source_layout
* dest_layout
* mean_values

> * scale_values

Reviewer 4: I think it makes sense to refer their description in `help` or provide their description here.

Author: This section was expected to cover only the functionality which present in MO Python API, but not in cli tool. So description of these params is expected to be in separate topics, which are common for python API and cli tool.

I am not sure how to add reference to help, as currently we have help generated automatically from this docstring: https://github.com/openvinotoolkit/openvino/blob/03a428f50c503b708be1f9a9bf54df99635e8013/tools/mo/openvino/tools/mo/convert.py#L84
